### PR TITLE
fixes #9649 feat(cirrus): add enrollment_status metric

### LINKED ISSUE #9649: Add support for EnrollmentStatus metric
https://mozilla-hub.atlassian.net/browse/EXP-3828

### PR DESCRIPTION
**Pull request by jeddai**
Because

- The enrollment_status metric has been accepted as a needed metric for experiment analysis

This commit

- Adds support for the enrollment_status metric into Cirrus

### CODE REVIEW

File: cirrus/server/Dockerfile
Comment (jeddai) [inline cirrus/server/Dockerfile:19]: This is here to have the build run with the change as it has been written in application-services here: https://github.com/mozilla/application-services/pull/5891

Before merging it will be removed

File: cirrus/server/cirrus/sdk.py
Comment (brennie) [inline cirrus/server/cirrus/sdk.py:16]: Sad type signature is sad 😭

File: cirrus/server/tests/test_telemetry.py
Comment (brennie) [inline cirrus/server/tests/test_telemetry.py:115]: I'm personally not a fan of destructuring with _ in python because it actually declares a variable named _ (unlike discarding immediately like in rust), but I'm not gonna block this pr on that.

File: cirrus/server/cirrus/sdk.py
Comment (brennie) [inline cirrus/server/cirrus/sdk.py:25]: It's kinda weird that the type is exposed as a member instead of a top level type. Is this for test stubbing?

**Review by brennie — APPROVED** [review]
No major issues, lgtm

File: cirrus/server/cirrus/sdk.py
Comment (jeddai) [reply cirrus/server/cirrus/sdk.py:16]: Yeah I was hoping there'd be a specific type for the metrics/pings but the return type of the Glean functions is listed as `Any`

File: cirrus/server/cirrus/sdk.py
Comment (jeddai) [reply cirrus/server/cirrus/sdk.py:25]: Nope, that's just kinda how Glean is used in Kotlin/Swift/Python as far as I'm aware

Comment (jeddai) [reply cirrus/server/cirrus/sdk.py:25]: Here's an example in iOS https://github.com/mozilla-mobile/firefox-ios/blob/8967b3f84d9a35f79c9b0b4cc1381a5926df4b4f/Client/Telemetry/SponsoredTileTelemetry.swift#L39C1-L45

Comment (brennie) [reply cirrus/server/cirrus/sdk.py:25]: wack.

**Review by yashikakhurana — CHANGES_REQUESTED** [review]
Thank you @jeddai overall looks great 🎉 🎉 , I am just curious don't we need data review for this?

**Comment by jeddai** [comment]
Yep! We do need a data review — going to be posting that shortly

**Comment by jeddai** [comment]
# Request for data collection review form

1) What questions will you answer with this data?
What are the enrollment statuses of all recipes a given user is enrolled in when recipes are finished being applied?

2) Why does Mozilla need to answer these questions? 
This data is required for furthering the development of Nimbus. Specifically, it will help data science resolve a current issue where \*"we get telemetry from users who do not enroll in deliveries that we expect them to"

\* ref. telemetry proposal document, linked in the ticket linked on this PR.

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?
None, no solution is sufficient unless the data collected in this metric is collected.
We chose to combine what would have been a significant number of new metrics into a single `enrollment_status` metric.

4) Can current instrumentation answer these questions?
No

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.   

<table>
  <tr>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>Enrollment Status</td>
    <td>Technical data</td>
    <td>https://mozilla-hub.atlassian.net/browse/EXP-3827</td>
  </tr>
</table>

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
The documentation can be found in the proposal document linked in the ticket above (internal only), the [metrics.yaml](https://github.com/mozilla/application-services/pull/5857/files#diff-ca3094e555a194d8c839be789f65c5434e9d04ad84ec363a3072cd89225d9a83) on this PR, and the [Glean dictionary](https://dictionary.telemetry.mozilla.org/).

7) How long will this data be collected?  Choose one of the following:
* I want to permanently monitor this data.
  * chumphreys@mozilla.com
  * project-nimbus@mozilla.com

8) What populations will you measure?
All populations that use the Nimbus SDK. All of Firefox Android, Firefox iOS, and ultimately all tools using the stateless Nimbus SDK as well.

9) If this data collection is default on, what is the opt-out mechanism for users?
Nimbus has internal tooling for opting out, it is up to the implementing applications to use this tooling correctly.

10) Please provide a general description of how you will analyze this data.
We'll analyze it in Jetstream and it will be used to create dashboards for experiment health as well as additional data.

11) Where do you intend to share the results of your analysis?
They will be shared internally to Mozilla through the Experimenter console as well as ad-hoc through queries against the raw Glean data.

12) Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection? If so:
No

data-review?@eliserichards

**Comment by eliserichards** [comment]
> # Request for data collection review form
> 
>     1. What questions will you answer with this data?
>        What are the enrollment statuses of all recipes a given user is enrolled in when recipes are finished being applied?
> 
>     2. Why does Mozilla need to answer these questions?
>        This data is required for furthering the development of Nimbus. Specifically, it will help data science resolve a current issue where *"we get telemetry from users who do not enroll in deliveries that we expect them to"
> 
> 
> * ref. telemetry proposal document, linked in the ticket linked on this PR.
> 
>     3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
>        None, no solution is sufficient unless the data collected in this metric is collected.
>        We chose to combine what would have been a significant number of new metrics into a single `enrollment_status` metric.
> 
>     4. Can current instrumentation answer these questions?
>        No
> 
>     5. List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.
> 
> 
> Measurement Description 	Data Collection Category 	Tracking Bug #
> Enrollment Status 	Technical data 	https://mozilla-hub.atlassian.net/browse/EXP-3827
> 
>     6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
>        The documentation can be found in the proposal document linked in the ticket above (internal only), the [metrics.yaml](https://github.com/mozilla/application-services/pull/5857/files#diff-ca3094e555a194d8c839be789f65c5434e9d04ad84ec363a3072cd89225d9a83) on this PR, and the [Glean dictionary](https://dictionary.telemetry.mozilla.org/).
> 
>     7. How long will this data be collected?  Choose one of the following:
> 
> 
>     * I want to permanently monitor this data.
>       
>       * [chumphreys@mozilla.com](mailto:chumphreys@mozilla.com)
>       * [project-nimbus@mozilla.com](mailto:project-nimbus@mozilla.com)
> 
> 
>     8. What populations will you measure?
>        All populations that use the Nimbus SDK. All of Firefox Android, Firefox iOS, and ultimately all tools using the stateless Nimbus SDK as well.
> 
>     9. If this data collection is default on, what is the opt-out mechanism for users?
>        Nimbus has internal tooling for opting out, it is up to the implementing applications to use this tooling correctly.
> 
>     10. Please provide a general description of how you will analyze this data.
>         We'll analyze it in Jetstream and it will be used to create dashboards for experiment health as well as additional data.
> 
>     11. Where do you intend to share the results of your analysis?
>         They will be shared internally to Mozilla through the Experimenter console as well as ad-hoc through queries against the raw Glean data.
> 
>     12. Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection? If so:
>         No
> 
> 
> data-review?@eliserichards

# Data Review Form (to be filled by Data Stewards)

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way? 

* Yes, through the Cirrus `metrics.yaml` file and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/)

2) Is there a control mechanism that allows the user to turn the data collection on and off? (Note, for data collection not needed for security purposes, Mozilla provides such a control mechanism) Provide details as to the control mechanism available.

* Yes, through the "Send Usage Data" preference in the application settings of the consumers.

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

* Yes, this data will be permanently collected. Charlie Humphreys (chumphreys@mozilla.com) will be monitoring as well as the Nimbus team (project-nimbus@mozilla.com).

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

* Category 1 - technical data.

5) Is the data collection request for default-on or default-off?

* Default on.

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?

* No. 

7) Is the data collection covered by the existing Firefox privacy notice? **If unsure: escalate to legal if:**

* Yes.

8) Does the data collection use a third-party collection tool? **If yes, escalate to legal.**

* No.

## Result

data-review+

**Comment by jaredlockhart** [comment]
Now we have a great opportunity.  Let's block this on https://github.com/mozilla/experimenter/pull/9658 and then rebase on top of it and then we'll be able to safely verify our first Cirrus change with integration tests 🎉 🎉 🎉 🎉

File: cirrus/server/tests/test_telemetry.py
Comment (jaredlockhart) [reply cirrus/server/tests/test_telemetry.py:115]: Yeah it's a pattern rather than an actual syntax in Python unfortunately.  The [docs](https://docs.python.org/3/reference/lexical_analysis.html#reserved-classes-of-identifiers) do refer to it but it's not explicitly supported as an actual syntax for unused variables outside of the wildcard in match statements.  But using `_` to denote an unused variable is a widely accepted pattern in Python so I think it's reasonable to use here.

Funnily, I found a rejected [PEP640](https://peps.python.org/pep-0640/) suggesting using `?` as an explicit unused variable identifier.  I can see why it was rejected.

**Review by jaredlockhart — APPROVED** [review]
Okay this all looks good to me thank you @jeddai ! 🎉   It would be nice as some future followup to actually be able to assert against the presence and behaviour of this telemetry in the integration tests cc @yashikakhurana 

Let's get this landed 🎉 🎉 🎉 🎉

**Comment by yashikakhurana** [comment]
> Okay this all looks good to me thank you @jeddai ! 🎉 It would be nice as some future followup to actually be able to assert against the presence and behaviour of this telemetry in the integration tests cc @yashikakhurana
> 
> Let's get this landed 🎉 🎉 🎉 🎉

yes we have next ticket for this -https://mozilla-hub.atlassian.net/browse/EXP-3814?atlOrigin=eyJpIjoiNWVlMjM5Yjk5ZTRhNGE4NWI1NDY5ODE0YzQzYmUxYjAiLCJwIjoiaiJ9 🎉

**Comment by jeddai** [comment]
@yashikakhurana I think you need to approve for the PR to merge

**Review by yashikakhurana — APPROVED** [review]
perfect 💯 . thank you @jeddai 🎉 🎉

**Pull request merged**
